From 67d2b7301c273dd099034f343a89179afc68a0be Mon Sep 17 00:00:00 2001
From: Sneh Bansal <snehb@codeaurora.org>
Date: Thu, 25 Jan 2018 12:17:37 +0530
Subject: [PATCH 09/13] Telephony: Send INITIAL_ATTACH only when it is
 applicable.

Issue is that SET_INITIAL_ATTACH is being sent whenever subId changes.

During bootup subId does change before records gets loaded. This results
in SET_INITIAL_ATTACH with incorrect APN setting/profile.

Solution is to send SET_INITIAL_ATTACH only when cdmaSubscriptionSource
is RUIM based and records have been loaded.

Change-Id: I68280db4831935326c732943e5b763d3b3a4c3fe

Fix to send initial attach apn

If there is no IA/default apn then do not perform initial
attach with any other available apns.

Change-Id: Ie3241747eb44df9070f2cb4874f13878f8ca3d9b
CRs-Fixed: 2178888
---
 .../internal/telephony/dataconnection/DcTracker.java  | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/java/com/android/internal/telephony/dataconnection/DcTracker.java b/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
index 39a1f073d..bda8b976e 100644
--- a/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
+++ b/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
@@ -443,7 +443,7 @@ public class DcTracker extends Handler {
     // member variables
     protected final Phone mPhone;
     private final UiccController mUiccController;
-    private final AtomicReference<IccRecords> mIccRecords = new AtomicReference<IccRecords>();
+    protected final AtomicReference<IccRecords> mIccRecords = new AtomicReference<IccRecords>();
     private DctConstants.Activity mActivity = DctConstants.Activity.NONE;
     private DctConstants.State mState = DctConstants.State.IDLE;
     private final Handler mDataConnectionTracker;
@@ -2095,6 +2095,11 @@ public class DcTracker extends Handler {
             }
         }
 
+        if ((iaApnSetting == null) && (defaultApnSetting == null) &&
+                !allowInitialAttachForOperator()) {
+            log("Abort Initial attach");
+            return;
+        }
         // The priority of apn candidates from highest to lowest is:
         //   1) APN_TYPE_IA (Initial Attach)
         //   2) mPreferredApn, i.e. the current preferred apn
@@ -2126,6 +2131,10 @@ public class DcTracker extends Handler {
         }
     }
 
+    protected boolean allowInitialAttachForOperator() {
+        return true;
+    }
+
     /**
      * Handles changes to the APN database.
      */
-- 
2.17.1

